### PR TITLE
[feat] Added random_apply to augmentaton containers

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
This PR added ```random_apply``` parameter to augmentation containers, which allows to apply augmentations in random order.

For ```PatchSequential```, this PR adds features on allowing true "patchwise" augmentations that patches in different images will apply different augmentations.

\cc @LinShiqi047

### CODE REVIEW

File: kornia/augmentation/container/image.py [hunk 1]
```diff
@@ -69,39 +81,99 @@ def __init__(
...
+
+    def _get_child_sequence(self) -> Iterator[Tuple[str, nn.Module]]:
+        if self.random_apply:
+            num_samples = int(torch.randint(*self.random_apply, (1,)).item())
+            indicies = torch.multinomial(
+                torch.ones((len(self),)), num_samples, replacement=True if num_samples > len(self) else False
+            )
+            return self._get_children_by_indicies(indicies)
+        else:
+            return self.named_children()
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:122]:
```suggestion
        return self.named_children()
```

File: kornia/augmentation/container/image.py [hunk 2]
```diff
@@ -69,39 +81,99 @@ def __init__(
...
+            self.random_apply = False
+
+    def _get_child_sequence(self) -> Iterator[Tuple[str, nn.Module]]:
+        if self.random_apply:
+            num_samples = int(torch.randint(*self.random_apply, (1,)).item())
+            indicies = torch.multinomial(
+                torch.ones((len(self),)), num_samples, replacement=True if num_samples > len(self) else False
+            )
+            return self._get_children_by_indicies(indicies)
+        else:
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:121]:
```suggestion
```
deepsource will complain later

File: kornia/augmentation/container/image.py [hunk 3]
```diff
@@ -69,39 +81,99 @@ def __init__(
...
+            return self.named_children()
+
+    def _get_children_by_indicies(self, indicies: torch.Tensor) -> Iterator[Tuple[str, nn.Module]]:
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:124]:
```suggestion
    def _get_children_by_indices(self, indices: torch.Tensor) -> Iterator[Tuple[str, nn.Module]]:
```

File: kornia/augmentation/container/image.py [hunk 4]
```diff
@@ -69,39 +81,99 @@ def __init__(
...
+        modules = list(self.named_children())
+        for idx in indicies:
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:126]:
```suggestion
        for idx in indices:
```

File: kornia/augmentation/container/augment.py [hunk 5]
```diff
@@ -25,6 +28,13 @@ class AugmentationSequential(ImageSequential):
             applied to each. If None, it will not overwrite the function-wise settings. Default: None.
         keepdim (bool, optional): whether to keep the output shape the same as input (True) or broadcast it
             to the batch form (False). If None, it will not overwrite the function-wise settings. Default: None.
+        random_apply(int, (int, int), optional): randomly select a sublist (order agnostic) of args to
```
Comment (edgarriba) [inline kornia/augmentation/container/augment.py:31]:
```suggestion
        random_apply: randomly select a sublist (order agnostic) of args to
```
we won't need to add typing by hand in docs anymore neither defaults

File: kornia/augmentation/container/video.py [hunk 6]
```diff
@@ -22,6 +25,12 @@ class VideoSequential(ImageSequential):
         *args (_AugmentationBase): a list of augmentation module.
         data_format (str): only BCTHW and BTCHW are supported. Default: BTCHW.
         same_on_frame (bool): apply the same transformation across the channel per frame. Default: True.
+        random_apply(int, (int, int), optional): randomly select a sublist (order agnostic) of args to
```
Comment (edgarriba) [inline kornia/augmentation/container/video.py:28]: maybe not for this PR, but as soon that all containers have the `random_apply` feature why not implementing it at base level ?

Comment (shijianjian) [reply kornia/augmentation/container/video.py:28]: The base level of containers is the ```ImageSequential``` class. This is implemented there but still needs some tweaks in different scenarios.

File: test/augmentation/test_container.py [hunk 7]
```diff
@@ -8,6 +9,30 @@
 from kornia.geometry.transform import bbox_to_mask
 
 
+def reproducibility_test(input, seq):
```
Comment (edgarriba) [inline test/augmentation/test_container.py:12]: if the test name does not follow `test_*` then pytest won't collect it

Comment (shijianjian) [reply test/augmentation/test_container.py:11]: It is alright. It is general function for performing reproducibility test. It will be called afterwards.

**Comment by shijianjian** [comment]
This is more or less ready to be merged. We may include this in 0.5.5. I will start another PR for enabling mix augmentation in containers.

**Review by edgarriba — APPROVED** [review]

**Pull request merged**